Monero-Peer-Subnet-Deduplication/code/Rust/src/main.rs

```diff
@@ -0,0 +1,201 @@
+use std::{
+    collections::HashSet,
+    convert::Infallible,
+    fs::OpenOptions,
+    io::Write,
+    net::SocketAddr,
+    sync::{LazyLock, OnceLock},
+    task::Poll,
+    time::Duration,
+};
+
+use dashmap::DashSet;
+use futures::{future::BoxFuture, stream, FutureExt};
+use tokio::{
+    sync::{mpsc, Semaphore},
+    time::timeout,
+};
+use tower::{make::Shared, util::MapErr, Service, ServiceExt};
+use tracing::error;
+use tracing_subscriber::{filter::LevelFilter, FmtSubscriber};
+
+use cuprate_p2p_core::{
+    client::{
+        handshaker::builder::{DummyCoreSyncSvc, DummyProtocolRequestHandler},
+        ConnectRequest, Connector, HandshakerBuilder, InternalPeerID,
+    },
+    services::{AddressBookRequest, AddressBookResponse},
+    BroadcastMessage, ClearNet, NetZoneAddress, Network, NetworkZone,
+};
+use cuprate_wire::{common::PeerSupportFlags, BasicNodeData};
+
+/// A set of all node's [`SocketAddr`] that we have successfully connected to.
+static SCANNED_NODES: LazyLock<DashSet<SocketAddr>> = LazyLock::new(DashSet::new);
+
+/// The [`Connector`] service to make outbound connections to nodes.
+static CONNECTOR: OnceLock<
+    Connector<
+        ClearNet,
+        AddressBookService,
+        DummyCoreSyncSvc,
+        MapErr<Shared<DummyProtocolRequestHandler>, fn(Infallible) -> tower::BoxError>,
+        fn(InternalPeerID<<ClearNet as NetworkZone>::Addr>) -> stream::Pending<BroadcastMessage>,
+    >,
+> = OnceLock::new();
+
+/// The channel that is used to communicate a successful connection.
+static BAD_PEERS_CHANNEL: OnceLock<mpsc::Sender<(SocketAddr, bool)>> = OnceLock::new();
+
+/// A [`Semaphore`] to limit the amount of concurrent connection attempts so we don't overrun ourself.
+static CONNECTION_SEMAPHORE: Semaphore = Semaphore::const_new(100);
+
+#[tokio::main(flavor = "multi_thread")]
+async fn main() {
+    FmtSubscriber::builder()
+        .with_max_level(LevelFilter::DEBUG)
+        .init();
+
+    let handshaker = HandshakerBuilder::<ClearNet>::new(BasicNodeData {
+        my_port: 0,
+        network_id: Network::Mainnet.network_id(),
+        peer_id: rand::random(),
+        support_flags: PeerSupportFlags::FLUFFY_BLOCKS,
+        rpc_port: 0,
+        rpc_credits_per_hash: 0,
+    })
+    .with_address_book(AddressBookService)
+    .build();
+
+    // Create and set the CONNECTOR global.
+    let connector = Connector::new(handshaker);
+    let _ = CONNECTOR.set(connector.clone());
+
+    // Create and set the BAD_PEERS_CHANNEL global.
+    let (bad_peers_tx, mut bad_peers_rx) = mpsc::channel(508);
+    BAD_PEERS_CHANNEL.set(bad_peers_tx).unwrap();
+
+    // start by connecting to the seed nodes
+    [
+        "176.9.0.187:18080",
+        "88.198.163.90:18080",
+        "66.85.74.134:18080",
+        "51.79.173.165:18080",
+        "192.99.8.110:18080",
+        "37.187.74.171:18080",
+        "77.172.183.193:18080",
```

Comment:
```suggestion
        "88.99.195.15:18080",
```
This was replaced
https://github.com/monero-project/monero/pull/9792

Reply:
@Boog900 , do you think this revision to the seed node list is a good idea? If so, @nahuhh can make a PR to `misc-research`.

Reply:
The other seeds will be enough. Although I am not against updating this and if you do I would recommend adding a comment that these seed nodes my change and that users should check if they are still in use by looking at monero-core so this doesn't have to be maintained in perpetuity.